### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by Sarveshgith**

### CODE REVIEW

File: Config/dbConnect.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
+
+const dbConnect = async() => {
```
Comment (ShyamSunder149) [inline Config/dbConnect.js]: Make sure you leave some proper spacing everywhere, Like in 
```suggestion
const dbConnect = async () => {
```
everywhere in the try catch block. Just an improvement for better readability

File: Constants/errcodes.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+exports.constants = {
+    VALIDATION_ERR : 400,
```
Comment (ShyamSunder149) [inline Constants/errcodes.js]: Check status code 422 and check whether you can use that instead of 400

File: Controllers/AdminController.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
+const AsyncHandler = require("express-async-handler");
+const bcrypt = require("bcrypt");
+const jwt = require("jsonwebtoken");
+const Admin = require("../Models/AdminModel");
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js:5]: have a single line space between npm packages and your packages. Again just better readability

File: Controllers/AdminController.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+const Book = require("../Models/BookModel");
+const Student = require("../Models/StudentModel");
+const DefinedError = require("../Middleware/DefinedError");
+const { errHandle } = require("../Middleware/errHandle");
+
+const RegisterUser = async (req, res) => {
+  try {
+    const { username, password, name } = req.body;
+
+    if (!username || !password || !name) {
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]:
```suggestion
    if ([username, password, name].some(field => !field)) {
```
Check Array.prototype.some docs

File: Controllers/AdminController.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+      throw new DefinedError(
+        400,
+        "error",
+        "Please Add All Fields",
+        "Admin Not Registered"
+      );
+    }
+
+    const UserExists = await Admin.findOne({ username });
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]:
```suggestion
    const user = await Admin.findOne({ username });
```
var name starts with small letter
and the name UserExists should have a boolean value but we have an object here so always its user

File: Controllers/AdminController.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+    if (UserExists) {
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]:
```suggestion
    if (user) {
```

File: Controllers/AdminController.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+      throw new DefinedError(
+        404,
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]: 404 is not found
use a better status code here

File: Controllers/AdminController.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        "error",
+        "Admin Already Exists",
+        "Admin Not Registered"
+      );
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    console.log(hashedPassword);
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]: remove logs for hashed password

File: Controllers/AdminController.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      res
+    );
+  }
+};
+
+const LoginUser = async (req, res) => {
+  try {
+    const { username, password } = req.body;
+
+    if (!username || !password) {
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]: array.prototype.some changes

File: Controllers/AdminController.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+      throw new DefinedError(
+        400,
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]: check status 422

File: Controllers/AdminController.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+          },
+        },
+        process.env.ACCESS_TOKEN_SECRET,
+        { expiresIn: "15m" }
+      );
+
+      return res.status(200).json({ accessToken });
+    }
+    throw new DefinedError(
+      401,
```
Comment (ShyamSunder149) [inline Controllers/AdminController.js]: check 403 status code

File: Controllers/BookController.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
+const AsyncHandler = require("express-async-handler");
+const Book = require("../Models/BookModel");
```
Comment (ShyamSunder149) [inline Controllers/BookController.js:3]: separate npm packages with yours

File: Controllers/BookController.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+const { errHandle } = require("../Middleware/errHandle");
+const DefinedError = require("../Middleware/DefinedError");
+
+const AddBook = async (req, res) => {
+  try {
+    console.log("Created Book: ", req.body);
+    const { title, author } = req.body;
+
+    if (!title || !author) {
```
Comment (ShyamSunder149) [inline Controllers/BookController.js]: array.prototype.some changes

File: Controllers/BookController.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    const book = await Book.create(req.body);
+    return res.status(200).json(book);
+  } catch (err) {
+    errHandle(err, err instanceof DefinedError, "Book Not Shelved", res);
+  }
+};
+
+// /api/book/Harry%20Potter
```
Comment (ShyamSunder149) [inline Controllers/BookController.js]: comment not needed

File: Controllers/BookController.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+const DelBook = async (req, res) => {
+  try {
+    const { title } = req.params;
```
Comment (ShyamSunder149) [inline Controllers/BookController.js]: delete book by ID not title. What if 2 books hold the same title?

File: Controllers/BorrowController.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
+const Borrow = require("../Models/BorrowModel");
+const Book = require("../Models/BookModel");
+const Student = require("../Models/StudentModel");
+const { DateDifference } = require("../utils/DateDiff");
+const DefinedError = require("../Middleware/DefinedError");
+const { errHandle } = require("../Middleware/errHandle");
+
+const DuePeriod = 14;
```
Comment (ShyamSunder149) [inline Controllers/BorrowController.js]: constants can be defined in a separate file and be used here. better coding practice

File: Controllers/BorrowController.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+const DueAmount = 10; // per day
+
+const BorrowBooks = async (req, res) => {
+  try {
+    const { regno, bookId } = req.body;
+
+    if (!regno || !bookId) {
```
Comment (ShyamSunder149) [inline Controllers/BorrowController.js]: array.prototype.some changes

File: Controllers/BorrowController.js [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  } catch (err) {
+    errHandle(err, err instanceof DefinedError, "Book Not Borrowed", res);
+  }
+};
+
+const ReturnBooks = async (req, res) => {
+  try {
+    const { regno, bookId } = req.body;
+
+    if (!regno || !bookId) {
```
Comment (ShyamSunder149) [inline Controllers/BorrowController.js]: array.prototype.some changes

File: Controllers/BorrowController.js [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+      );
+    }
+
+    const borrowedAt = new Date(borrowedRecord.borrowedAt);
+    const currentDate = new Date();
+    const diff = DateDifference(currentDate, borrowedAt);
+
+    console.log(`It's been ${diff} days since the book was borrowed!`);
+
+    let due = 0;
```
Comment (ShyamSunder149) [inline Controllers/BorrowController.js]: have these kinda calculations in the util classes and call the method here. Don't bring up the logical funcs in controllers

File: Controllers/StudentController.js [hunk 20]
```diff
@@ -0,0 +1,129 @@
+const AsyncHandler = require("express-async-handler");
+const bcrypt = require("bcrypt");
+const jwt = require("jsonwebtoken");
+const Student = require("../Models/StudentModel");
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js:5]: split npm packages with your packages. Kindly make this change every where in these files

File: Controllers/StudentController.js [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+const Book = require("../Models/BookModel");
+const { errHandle } = require("../Middleware/errHandle");
+const DefinedError = require("../Middleware/DefinedError");
+
+const RegisterUser = async (req, res) => {
+  try {
+    const { regno, password, name } = req.body;
+
+    if (!regno || !password || !name) {
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js]: array.prototype.some changes. make this change everywhere in these files

File: Controllers/StudentController.js [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+      throw new DefinedError(
+        400,
+        "error",
+        "User Not Added",
+        "User Not Registered"
+      );
+    }
+
+    const UserExists = await Student.findOne({ regno });
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js:23]: same change as suggested above

File: Controllers/StudentController.js [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+      throw new DefinedError(
+        404,
+        "error",
+        "User Not Found",
+        "User Not Registered"
+      );
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    console.log(hashedPassword);
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js]: remove log

File: Controllers/StudentController.js [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+          },
+        },
+        process.env.ACCESS_TOKEN_SECRET,
+        { expiresIn: "15m" }
+      );
+
+      return res.status(200).json({ accessToken });
+    } else {
+      throw new DefinedError(
+        401,
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js]: status 403

File: Controllers/StudentController.js [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+        "User Cannot Login"
+      );
+    }
+  } catch (err) {
+    errHandle(err, err instanceof DefinedError, "User Cannot Login", res);
+  }
+};
+
+const GetBooks = async (req, res) => {
+  try {
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js]: try keeping the same end point for getbooks and getbook. If params is not passed then getbooks will execute
if there are any params then getbook according to that param. By this way there wont be 2 functions and you can add any type of filters with those params

File: Routes/StudentRoute.js [hunk 26]
```diff
@@ -0,0 +1,11 @@
+const express = require("express");
+const { RegisterUser, LoginUser, GetBooks, GetBook } = require("../Controllers/StudentController");
+const router = express.Router();
+const ValidToken = require("../Middleware/Auth");
+
+router.post("/register", RegisterUser);
+router.post("/login", LoginUser);
+router.get("/books", ValidToken, GetBooks);
+router.get("/books/search", ValidToken, GetBook);
+//api/student/books/search?author=Colleen%20Hoover
```
Comment (ShyamSunder149) [inline Routes/StudentRoute.js]: remove comments

File: utils/DateDiff.js [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+    const dateM2 = new Date(date2).getTime();
+
+    const diffM = Math.abs(dateM2 - dateM1);
+
+    const diffday = Math.ceil(diffM / (1000 * 60 * 60 * 24));
+
+    return diffday;
+};
+
+// const date1 = "2024-06-01";
```
Comment (ShyamSunder149) [inline utils/DateDiff.js]: remove commented code

**Review by ShyamSunder149 — COMMENTED** [review]
Incorporate suggested changes in your code. Try using chat gpt if you feel like the code can be improved somewhere. Focus on formatting and status codes. And try making changes I suggested in all the parts of the code eventhough I missed to mark it somewhere

File: utils/DateDiff.js [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+    const DueAmount = 10;
+    let due = 0;
+    const dateM1 = new Date(date1).getTime();
+    const dateM2 = new Date(date2).getTime();
+
+    const diffM = Math.abs(dateM2 - dateM1);
+
+    const diffday = Math.ceil(diffM / (1000 * 60 * 60 * 24));
+
+    if(diffday > DuePeriod){
```
Comment (ShyamSunder149) [inline utils/DateDiff.js]:
```suggestion
    return diffday > DuePeriod ? (diffday - DuePeriod) * DueAmount : 0
```

File: utils/DateDiff.js [hunk 29]
```diff
@@ -0,0 +1,19 @@
+function DueAmtCalc(date1, date2){
+
+    const DuePeriod = 14;
+    const DueAmount = 10;
+    let due = 0;
```
Comment (ShyamSunder149) [inline utils/DateDiff.js]: remove this

File: Routes/BookRoute.js [hunk 30]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
+const { AddBook, DelBook } = require("../Controllers/BookController");
+const router = express.Router();
+
+//router.use(ValidToken);
```
Comment (ShyamSunder149) [inline Routes/BookRoute.js:5]: remove comments

File: Middleware/Auth.js [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+      next();
+    });
+  } else {
+    res.status(403);
+    throw new Error("User not authorized or token is missing");
+  }
+});
+
+const AuthorizeUser = (req, res, next) => {
+  if (req.user && res.user.id === req.params.userId) next();
```
Comment (ShyamSunder149) [inline Middleware/Auth.js:31]: improvise the if conditions by always handling the failure case first. In this scenario you can try this
```suggestion
 if(!req.user) // do something
 if(res.user.id != req.params.userId) // do something
 next()
```
by this way you are breaking down the complexity of the code and code will be easy to read.
Try to implement this in all places in code

File: Controllers/StudentController.js [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+        "Invalid Credantials",
+        "User Cannot Login"
+      );
+    }
+  } catch (err) {
+    errHandle(err, err instanceof DefinedError, "User Cannot Login", res);
+  }
+};
+
+const GetBook = async (req, res) => {
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js:92]: try to integrate the with GetBooks endpoint. If a param is passed then the book should be filtered accordingly in the book controller itself

**Comment by Sarveshgith** [comment]
Resolved!

File: utils/DateDiff.js [hunk 33]
```diff
@@ -1,19 +1,18 @@
...
-    let due = 0;
-    const dateM1 = new Date(date1).getTime();
-    const dateM2 = new Date(date2).getTime();
+  const diffM = Math.abs(dateM2 - dateM1);
 
-    const diffM = Math.abs(dateM2 - dateM1);
+  const diffday = Math.ceil(diffM / (1000 * 60 * 60 * 24));
 
-    const diffday = Math.ceil(diffM / (1000 * 60 * 60 * 24));
+  if (diffday > DuePeriod) {
```
Comment (ShyamSunder149) [inline utils/DateDiff.js:12]: this can be changed to ternary operator

File: Constants/errcodes.js [hunk 34]
```diff
@@ -1,7 +1,7 @@
 exports.constants = {
-    VALIDATION_ERR : 400,
-    UNAUTHORIZED : 401,
-    FORBIDDEN : 403,
-    NOT_FOUND :  404,
-    SERVER_ERROR: 500
-}
\ No newline at end of file
+  VALIDATION_ERR: 422,
```
Comment (ShyamSunder149) [inline Constants/errcodes.js:2]: can we handle err 409 here?

File: Controllers/StudentController.js [hunk 35]
```diff
@@ -94,15 +94,15 @@ const GetBook = async (req, res) => {
 
   try {
     let book;
-    if(author || title){
+    if (author || title) {
       const query = {};
       if (title) query.title = new RegExp(title, "i");
       if (author) query.author = new RegExp(author, "i");
-        book = await Book.find({
+      book = await Book.find({
```
Comment (ShyamSunder149) [inline Controllers/StudentController.js:101]: Can you check is there any way to efficiently handle the or case in terms of code?

**Review by ShyamSunder149 — COMMENTED** [review]
Just some more clarifications needed

**Pull request merged**